File: Explore Card Track/4. Array and String/0014-longest-common-prefix/0014-longest-common-prefix.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/**
 * @brief Finds the longest common prefix among an array of strings.
 *
 * @param strs (char**): Array of input strings.
 * @param strsSize (int): Number of strings in the array.
 *
 * @return (char*): The longest common prefix, or an empty string if none exists.
 */
char* longestCommonPrefix(char** strs, int strsSize) {
    // Edge case: if there are no strings, return an empty prefix
    if (strsSize == 0) {
        return "";
    }
    
    // Initialize the prefix with the first string
    char* prefix = strs[0];
    int prefix_len = strlen(prefix);
    
    // Loop through each string in the array, starting from the second string
    for (int i = 1; i < strsSize; i++) {
        int j = 0;
        
        // Compare characters until we find a mismatch or reach the end of one string
        while (j < prefix_len && strs[i][j] == prefix[j]) {
            j++;
        }
        
        // Update the prefix length based on the matched characters
        prefix_len = j;
        
        // If at any point the common prefix becomes empty, return immediately
        if (prefix_len == 0) {
            return "";
        }
    }
    
    // Allocate memory for the result and copy the prefix up to the matched length
    char* result = (char*)malloc((prefix_len + 1) * sizeof(char));
    strncpy(result, prefix, prefix_len);
    result[prefix_len] = '\0'; // Null-terminate the string
    
    return result;
}

File: Explore Card Track/4. Array and String/0014-longest-common-prefix/0014-longest-common-prefix.py
from typing import List

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        """
        Finds the longest common prefix among an array of strings.

        Parameters:
        strs (List[str]): List of input strings.

        Returns:
        str: The longest common prefix, or an empty string if none exists.
        """
        
        # Edge case: If the input list is empty, return an empty string
        if not strs:
            return ""
        
        # Start with the first string as the initial prefix
        prefix = strs[0]
        
        # Compare prefix with each string in the list
        for string in strs[1:]:
            # Reduce the prefix size until it matches the start of `string`
            while string[:len(prefix)] != prefix:
                prefix = prefix[:-1]  # Shorten the prefix by removing the last character
                if not prefix:
                    return ""  # No common prefix exists
        
        return prefix
